File: server/Services/PcBuildService.cs
namespace PCpals.Services;

public class PcBuildService{
    private readonly PcBuildRepository repo;
    public PcBuildService(PcBuildRepository repo){
        this.repo = repo;
    }

    internal PcBuild CreatePcBuild(PcBuild pcBuildData, string userId){
        if(userId == null)throw new Exception("Not Authorized");
        PcBuild pcBuild = repo.CreatePcBuild(pcBuildData);
        return pcBuild;
    }

    internal List<PcBuild> GetPersonalBuilds(string userId){
        if(userId == null)throw new Exception("No user found");
        List<PcBuild> pcBuilds = repo.GetPersonalBuilds(userId);
        return pcBuilds;
    }

    internal PcBuild GetPcById(int pcId){
        PcBuild pcBuild = repo.GetPcById(pcId);
        return pcBuild;
    }

    internal string DeletePcBuild(int pcBuildId, string userId){
        if(userId == null)throw new Exception("Not Authorized");
        repo.DeletePcBuild(pcBuildId);
        string message = "Build deleted.";
        return message;
    }

    internal PcBuild UpdateBuildParts(PcBuild updateBuildData ,int pcId,string userId){
        PcBuild originalPc = GetPcById(pcId);
        if(originalPc == null)throw new Exception("PC ID is invalid.");
        else if(originalPc.CreatorId == userId){
        originalPc.PcCase = updateBuildData.PcCase?.Length > 0 ? updateBuildData.PcCase : originalPc.PcCase;
        originalPc.CasePicture = updateBuildData.CasePicture?.Length > 0 ? updateBuildData.CasePicture : originalPc.CasePicture;
        originalPc.PcCpu = updateBuildData.PcCpu?.Length > 0 ? updateBuildData.PcCpu : originalPc.PcCpu;
        originalPc.Gpu = updateBuildData.Gpu?.Length > 0 ? updateBuildData.Gpu : originalPc.Gpu;
        originalPc.MotherBoard = updateBuildData.MotherBoard?.Length > 0 ? updateBuildData.MotherBoard : originalPc.MotherBoard;
        originalPc.Ram = updateBuildData.Ram?.Length > 0 ? updateBuildData.Ram : originalPc.Ram;
        originalPc.PcStorage = updateBuildData.PcStorage?.Length > 0 ? updateBuildData.PcStorage : originalPc.PcStorage;
        originalPc.PowerSupply = updateBuildData.PowerSupply?.Length > 0 ? updateBuildData.PowerSupply : originalPc.PowerSupply;
        PcBuild newPcBuild = repo.UpdateBuildParts(originalPc);
        return newPcBuild;
        }else{throw new Exception("Not Authorized");}
    }

    internal PcBuild UpdateShare(PcBuild buildData,int pcId, string userId){
        PcBuild originalPc = GetPcById(pcId);
        if(originalPc.CreatorId == userId){
            originalPc.IsPrivate = buildData.IsPrivate != originalPc.IsPrivate ? buildData.IsPrivate : originalPc.IsPrivate;
            PcBuild newBuild = repo.UpdateShare(originalPc);
            return newBuild;
        }else{throw new Exception("Not Authorized");}
    }

    internal PcBuild UpdatePowerScore(PcBuild updateData, int pcId, string userId){
        PcBuild originalPc = GetPcById(pcId);
        if(originalPc == null)throw new Exception("No PC found with that Id.");
        if(originalPc.CreatorId == userId){
            originalPc.PowerScore = updateData.PowerScore;
            originalPc.Price = updateData.Price;
            originalPc.Watts = updateData.Watts;
            originalPc.MaxWattage = updateData.MaxWattage;
            PcBuild updatedData = repo.UpdatePowerScore(originalPc);
            return updatedData;
        }else{throw new Exception("Not Authorized.");}
    }

    internal List<PcBuild> GetAllSharedBuilds(){
        List<PcBuild> pcBuilds = repo.GetAllSharedBuilds();
        return pcBuilds;
    }

    internal PcBuild UpdateRating(PcBuild updateData, int buildId, string userId){
        PcBuild originalBuild = GetPcById(buildId);
        if(originalBuild == null)throw new Exception("No Pc found with that Id");
        originalBuild.Rating = updateData.Rating;
        PcBuild updatedBuild = repo.UpdateRating(originalBuild);
        return updatedBuild;
    }

}

File: client/src/services/PcBuildService.js

import { AppState } from "../AppState"
import { PcBuild } from "../models/PcBuild"
import Pop from "../utils/Pop"
import { api } from "./AxiosService"
import { pcPartService } from "./pcPartService"


class PcBuildService{

    async createBuild(buildData){
        buildData.isPrivate = true
        let response = await api.post('api/pcBuilds', buildData)
        let build = new PcBuild(response.data)
        AppState.userBuilds.unshift(build)
        return build
    }
    async getUserBuilds(){
        let response = await api.get('api/pcBuilds/account')
        let userBuilds = response.data.map(build => new PcBuild(build))
        AppState.userBuilds = userBuilds
        // console.log(userBuilds)
    }

    async getPcById(pcId){
        let response = await api.get(`api/pcBuilds/${pcId}`)
        AppState.activeBuildToEdit = new PcBuild(response.data)
    }

    async deleteBuild(buildId){
        let response = await api.delete(`api/pcBuilds/${buildId}`)
        let buildIndex = AppState.userBuilds.findIndex(build => build.id == buildId)
        AppState.userBuilds.splice(buildIndex, 1)
        return response.data
    }

    async updateBuildParts(pcId, stockPartData){
        let buildData = {}
        if(stockPartData.type == 'case'){
            buildData.pcCase = stockPartData.name
            buildData.casePicture = stockPartData.productImage
        }else if(stockPartData.type == 'cpu'){ buildData.pcCpu = stockPartData.name}
        else if(stockPartData.type == 'gpu'){ buildData.gpu = stockPartData.name}
        else if(stockPartData.type == 'motherB'){buildData.motherBoard = stockPartData.name}
        else if(stockPartData.type == 'ram'){buildData.ram = stockPartData.name}
        else if(stockPartData.type == 'storage'){buildData.pcStorage = stockPartData.name}
        else if(stockPartData.type == 'powerS'){
            buildData.powerSupply = stockPartData.name
        }
        else if(stockPartData.type == 'cooler'){buildData.cooler = stockPartData.name}
        let response = await api.put(`api/pcBuilds/${pcId}/parts`, buildData)
        let updatedBuild = new PcBuild(response.data)
        AppState.activeBuildToEdit = updatedBuild
    }

    async updateShare(pcId){
        let activeBuild = AppState.userBuilds.find(build => build.id == pcId)
        let hasAllParts = await this.checkForParts(pcId)
        if(activeBuild.isPrivate == true && hasAllParts == true){
            activeBuild.isPrivate = false
            let updateData = {isPrivate: activeBuild.isPrivate}
            let response = await api.put(`api/pcBuilds/${pcId}/share/update`, updateData)
            let newBuildData = new PcBuild(response.data)
            AppState.userBuilds = await AppState.userBuilds.map(build => build.id !== pcId ? build : newBuildData)
            Pop.success('PC was shared for others to see!')
        }else if(activeBuild.isPrivate == true && hasAllParts == false){
            Pop.error('PC does not have all required parts.')
        }
        else if(activeBuild.isPrivate == false){
            activeBuild.isPrivate = true
            let updateData = {isPrivate: activeBuild.isPrivate}
            let response = await api.put(`api/pcBuilds/${pcId}/share/update`, updateData)
            let newBuildData = new PcBuild(response.data)
            AppState.userBuilds = AppState.userBuilds.map(build => build.id !== pcId ? build : newBuildData)
            Pop.success('PC was made private.')
        }
    }

    async updatePowerScore(pcId, average, price, wattage, maxWatts){
        let buildData = {powerScore: average, price: price, watts: wattage, maxWattage: maxWatts}
        // console.log(buildData)
        let response = await api.put(`api/pcBuilds/${pcId}/powerScore`, buildData)
        let newBuild = new PcBuild(response.data)
        AppState.activeBuildToEdit = newBuild
    }

    async updateRating(ratingData, buildId){
        let buildData = {rating: ratingData}
        // console.log(buildData, buildId)
        let response = await api.put(`api/pcBuilds/${buildId}/rating`, buildData)
        let updatedBuild = new PcBuild(response.data)
        AppState.activeBuildForReview = updatedBuild
        AppState.recentBuilds = AppState.recentBuilds.map(pc => pc.id !== buildId ? pc : updatedBuild)
    }

    async getSharedBuilds(){
        let response = await api.get('api/pcBuilds/shared')
        let allBuilds = await response.data.map(build => new PcBuild(build))
        AppState.recentBuilds = allBuilds
    }
    
    async checkForParts(buildId){
        let hasParts = false
        await pcPartService.getBuildParts(buildId)
        await pcPartService.allClear()
        let partCount = AppState.activeBuildParts.length
        if(partCount >= 7){
            hasParts = true
        }
        return hasParts
    }

}

export const pcBuildService = new PcBuildService()


